### PR TITLE
codeql: use native operators for sdk/osmo numerical values 

### PR DESCRIPTION
**Pull request by pysel**
Closes: #5963

## What is the purpose of the change

Adds a codeql pack named `osmoql` to the repo with a codeql query that looks for binary operators being used between numerical sdk/osmo types, that should use native methods.

Example:
```
sdk.Int{1} == sdk.Int{1} // wrong (most cases)
sdk.Int{1}.Equals(sdk.Int{1}) // correct
```

## Current problems

* Autobuild step in codeql action does not include _test.go files in the database, hence, `check-sdk-operators.ql` query won't catch errors in test files (for now)
* Same step for some reason ignores submodules' contents (eg: osmomath, osmoutils packages) 

## Testing and Verifying

[manually](https://github.com/pysel/osmosis/pull/67)

## Where is the change documented? 

changelog

### CODE REVIEW

**Comment by pysel** [comment]
devbot help

**Comment by devbot-wizard** [comment]
Hi! I'm DevBot, a bot that helps with common tasks in the development process. Commands:
- `devbot fix`: Fix basic errors in the PR. (e.g. imports, changelog conflicts, go mod conflicts)
- `devbot merge base`: Merge the base branch into the PR branch.
- `devbot add changelog [misc/feat/bug/api/sec] [message]`: Add a changelog entry to the PR. (e.g. `devbot add changelog feat Added a new feature`)
    - If message is blank, defaults to PR title.
- `devbot re-pr`: Re-PR the PR, useful for external contributors where we have no edit perms.

File: .github/osmoql/src/check-sdk-operators.ql
Comment (pysel) [inline .github/osmoql/src/check-sdk-operators.ql:17]: note: actually, we only need to check for `==` operator, because other operators are not defined on go structs, which will raise an error in the editor either ways

Comment (ValarDragon) [reply .github/osmoql/src/check-sdk-operators.ql:17]: Lets just check equals then?

**Review by ValarDragon — APPROVED** [review]
Great job! Can you add the git diff conditions we use for our go tests, to make this not over-activate?

**Comment by ValarDragon** [comment]
I'm not seeing a CI run, can you make a relevant change so we can see if this works?

**Comment by pysel** [comment]
hey @ValarDragon! It is not shown in CI because it only gets [activated](https://github.com/osmosis-labs/osmosis/pull/6267/files#diff-10a858a11563c34627198618ac039cb83e2bf54ac2847b66158d7358339dbcb5R3-R12) when `*.go` paths are modified in the pull request

**Comment by ValarDragon** [comment]
Can you add this into a test somewhere so we can see a run?

**Comment by ValarDragon** [comment]
This doesn't seem to be working, right? CodeQL normally leaves a github comment for errors (at least in the cosmos SDK)

**Comment by pysel** [comment]
> This doesn't seem to be working, right? CodeQL normally leaves a github comment for errors (at least in the cosmos SDK)

yeah, but I think it is something with repo's security parameters. the same workflow works well on my [fork](https://github.com/pysel/osmosis/pull/67), I will check the details of how codeql works and figure out what's the problem here

File: .github/osmoql/src/check-sdk-operators.ql
Comment (pysel) [reply .github/osmoql/src/check-sdk-operators.ql:16]: tldr, there is no need to check other operators

File: .github/osmoql/codeql-pack.lock.yml
Comment (pysel) [reply .github/osmoql/codeql-pack.lock.yml:5]: you mean this file? if so, yes, `qlpack.yml` and `codeql-pack.lock.yml` are similar to `buf.yml` and `buf.lock` in a sense

**Comment by mattverse** [comment]
@pysel sounds like a great idea! Let's go with approved review

File: .github/osmoql/codeql-pack.lock.yml
Comment (pysel) [reply .github/osmoql/codeql-pack.lock.yml:5]: step to generate - run `codeql pack install`

**Comment by pysel** [comment]
I am pretty sure this workflow does not show any warnings because the action is not on main (won't be shown in CI until in default branch). I think we can merge this PR and I will open a new "demo" PR to check wether it works. 

Code scanning alerts are set correctly in the repo, as far as I can tell, so I think the error is in the fact that the workflow is not in main.

cc: @ValarDragon @mattverse

**Comment by p0mvn** [comment]
> to reviewers: CodeQL analyze action is very time-expensive (around 25 mins to execute). wdyt about changing the trigger of this workflow from a simple push to (for example) an approved review?
> 
> cc: @ValarDragon @mattverse

Would it then block merge after approval?

I don't think we should be introducing this bottleneck. How about we run this only on merge to main?

**Comment by p0mvn** [comment]
@pysel could you please change this to run once per day on schedule and create an issue if breakages found?

If an issue already exists, close the old one. Thanks

**Comment by pysel** [comment]
> > to reviewers: CodeQL analyze action is very time-expensive (around 25 mins to execute). wdyt about changing the trigger of this workflow from a simple push to (for example) an approved review?
> > cc: @ValarDragon @mattverse
> 
> Would it then block merge after approval?
> 
> I don't think we should be introducing this bottleneck. How about we run this only on merge to main?

are you referring to the same merge block as we currently have with auto go mod upgrade workflow? if so, it will not block merge, since the reason of block in go mod case is push made with GITHUB_TOKEN creds

**Comment by pysel** [comment]
hey @p0mvn, just bumping this, should we go with a manual, scheduled, per-PR approach for running this workflow after all?

also, https://github.com/osmosis-labs/osmosis/pull/6267#issuecomment-1732647668

**Pull request closed without merging**
